docs/design-documents/20240430-schema-support.md

```diff
@@ -0,0 +1,413 @@
+<h1>Schema support</h1>
+
+<!-- TOC -->
+  * [The goal](#the-goal)
+  * [Requirements](#requirements)
+  * [Schema structure](#schema-structure)
+  * [Schema operations](#schema-operations)
+  * [Implementation](#implementation)
+    * [Schema storage](#schema-storage)
+      * [Option 1: Conduit itself hosts the schema registry](#option-1-conduit-itself-hosts-the-schema-registry)
+      * [Option 2: A centralized, external schema registry, accessed through Conduit](#option-2-a-centralized-external-schema-registry-accessed-through-conduit)
+      * [Option 3: A centralized, external schema registry, accessed by connectors directly](#option-3-a-centralized-external-schema-registry-accessed-by-connectors-directly)
+      * [Chosen option](#chosen-option)
+    * [Schema format](#schema-format)
+      * [Internal schema format](#internal-schema-format)
+        * [Option 1: Avro](#option-1-avro)
+        * [Option 2: Protobuf schema](#option-2-protobuf-schema)
+      * [Chosen option](#chosen-option-1)
+    * [Schema service interface](#schema-service-interface)
+      * [Option 1: Stream of commands and responses](#option-1-stream-of-commands-and-responses)
+      * [Option 2: Exposing a gRPC service in Conduit](#option-2-exposing-a-grpc-service-in-conduit)
+      * [Chosen option](#chosen-option-2)
+  * [Required changes](#required-changes)
+    * [Conduit](#conduit)
+    * [Conduit Commons](#conduit-commons)
+    * [Connector SDK](#connector-sdk)
+    * [Processor SDK](#processor-sdk)
+  * [Summary](#summary)
+  * [Other considerations](#other-considerations)
+<!-- TOC -->
+
+## The goal
+
+The goal of schema support is to make the information about data types available
+in a pipeline. This is needed to cover a few use cases, such as automatically
+creating a destination collection or propagating schema changes.
+
+## Requirements
+
+1. Records **should not** carry the whole schema.
+
+   Reason: If a record would carry the whole schema, that might increase a
+   record's size significantly.
+2. Sources and destinations need to be able to work with multiple schemas.
+   
+   Reason: Multiple collections support.
+3. A schema should be accessible across pipelines and Conduit instances.
+
+   Reason: Pipelines might work together to stream data from A to B, through an
+   intermediary C. For example, PostgreSQL to Kafka, and then Kafka to
+   Snowflake. Also, in the future, pipelines might run in isolation on different
+   instances. Additionally, Conduit may be used so that there's one pipeline per
+   instance. In such cases, we'd need this feature.
+4. It should be possible for a schema to evolve.
```

Comment:
Are we addressing this requirement? I can't find anything about schema versioning, which could probably address this part.

Reply:
From our discussions, it didn't seem that we had a use case for working with schema versions directly. The purpose of this requirement is just to highlight that a schema might evolve, and that a pipeline should be able to handle it (e.g. schema is not overwritten, older schemas can still be fetched with their IDs, and so on). I can rephrase this requirement to clarify it.

Reply:
I'm thinking of how will the destination connector figure out the schema evolved? If the schema could be identified using a name + version it could store in memory the last version it has seen for a schema name and then trigger some action if it sees a new version. Without versions, the connector would have no idea if the new schema in a record is actually an older one or a new one.

Reply:
It looks like the schema version's ID should be enough for that, or maybe no?

Reply:
The one that came after.:) A record that came after and has a different schema ID signals that its schema is a newer one.

Reply:
That's _probably_ the case, but might not always be the case, depending on how the source connector reads records. It could very well be that it gets some late-arriving records with the old schema. Or imagine two source connectors reading from independent sources, which have the same record schema. The schema gets updated in one source, but not the other, then records with different schema IDs might be interleaved for a while.

In any case, I don't want us making assumptions, rather give the destination the info it needs to make the right decision.

Reply:
I believe it would be interesting to discuss such cases a bit deeper, but given that it's just one field (the version field) that's readily available, it's probably easiest just to expose it and, as you said, give the destination more info. :+1: